### PR TITLE
Add setters to the StashNotifier.DescriptorImpl class

### PR DESCRIPTION
**Pull request by alexchiri**
In order to be able to configure this plugin using the [Configuration As Code plugin](https://github.com/jenkinsci/configuration-as-code-plugin), the StashNotifier.DescriptorImpl needs to have setters or public fields. Using YAML to configure Jenkins is pretty nice and this would enable this to work with this plugin as well.

### CODE REVIEW

**Comment by scaytrase** [comment]
All or most of this properties could be set while construction. While should we allow dynamically change the settings during lifecycle. Isn't it enough to pass YAML configuration to constructor in some way?

**Comment by alexchiri** [comment]
@scaytrase Unfortunately, at least as far as [I can see in the plugin implementation](https://github.com/jenkinsci/configuration-as-code-plugin/blob/master/plugin/src/main/java/io/jenkins/plugins/casc/BaseConfigurator.java#L50), it works by discovering fields and setters to determine what properties can be set from the YAML. They have some guidelines for plugin developers [here](https://github.com/jenkinsci/configuration-as-code-plugin/blob/master/PLUGINS.md).  There can be some custom implementation be done on the configuration-as-code plugin side to handle different ways of configuring plugins, but I felt since stash notifier has quite a basic config, that would be a bit of an overkill.

File: src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java
Comment (darxriggs) [inline src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java]: Please use correct indentation for all new code.
Here and in other places spaces & tabs are currently mixed up.

File: src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java
Comment (darxriggs) [inline src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java]: Please annotate all setters with `@DataboundSetter`.
See **Step 1** in https://github.com/jenkinsci/configuration-as-code-plugin/blob/configuration-as-code-1.8/docs/PLUGINS.md.

**Review by darxriggs — COMMENTED** [review]
Please also use data-binding via `request.bindJson(this, json)` inside `configure(...)`.
See **Step 5** in https://github.com/jenkinsci/configuration-as-code-plugin/blob/configuration-as-code-1.8/docs/PLUGINS.md.

There should also be two tests - one for configuring Jenkins via YAML and one for exporting a configured Jenkins to YAML.
See **How to test?** in https://github.com/jenkinsci/configuration-as-code-plugin/blob/configuration-as-code-1.8/docs/PLUGINS.md.

Note: I am not a maintainer of this plugin. But I already have migrated some other plugins to be compatible with JCasC.

**Comment by alexchiri** [comment]
@darxriggs Thank you for the feedback and review. Will work further on the PR soon.

**Comment by scaytrase** [comment]
@alexchiri as soon I have no views on architectural future of this plugin version (only some thoughts on 2.x version), I think we can accept this PR after suggested impovements

File: pom.xml
Comment (darxriggs) [inline pom.xml:14]: I suggest  to reduce it to the following snippet:
```
<properties>
    <revision>1.18</revision>
    <changelist>-SNAPSHOT</changelist>
    <jenkins.version>2.60.3</jenkins.version>
    <java.level>8</java.level>
</properties>
```
The rest is already covered by the parent pom nowadays.
(This can be verified with `mvn help:effective-pom`.)

File: pom.xml
Comment (darxriggs) [inline pom.xml]: If you like, there is already `3.40` - see https://github.com/jenkinsci/plugin-pom/releases.

File: src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java
Comment (darxriggs) [inline src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java:677]: It seems you missed a `@DataBoundSetter` here.

File: src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java
Comment (darxriggs) [inline src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java]: For Strings like this one I suggest to use something like Apache Commons Lang [StringUtils::trimToNull](https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/StringUtils.html#trimToNull-java.lang.String-) or Jenkins [Util::fixEmptyAndTrim](https://javadoc.jenkins.io/hudson/Util.html#fixEmptyAndTrim-java.lang.String-). This improves user experience by better handling strings with leading/trailing spaces.

File: src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java
Comment (darxriggs) [inline src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java]: See comment about `setStashRootUrl()` on trimming.

File: src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java
Comment (darxriggs) [inline src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java]: See comment about `setStashRootUrl()` on trimming.

File: src/test/java/org/jenkinsci/plugins/stashNotifier/ConfigAsCodeTest.java
Comment (darxriggs) [inline src/test/java/org/jenkinsci/plugins/stashNotifier/ConfigAsCodeTest.java:35]: Verifying `stashRootUrl` is missing here.

File: src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java
Comment (darxriggs) [inline src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java]: I suggest to keep the order consistent for easier reading/comparison:
* field declarations in `StashNotifier` (maybe)
* setters/getters in `StashNotifier` (maybe)
* default values (lines 751-758) in `StashNotifier::configure`
* assertThat calls (lines 27-33) in `ConfigAsCodeTest::should_support_jcasc_from_yaml`
* setter calls (lines 40-47) in `ConfigAsCodeTest::should_support_jcasc_to_yaml`

File: src/test/resources/org/jenkinsci/plugins/stashNotifier/ConfigAsCodeTest/configuration-as-code.yml
Comment (darxriggs) [inline src/test/resources/org/jenkinsci/plugins/stashNotifier/ConfigAsCodeTest/configuration-as-code.yml]: Even if it's only an example value, I suggest to use something like `https://my.company.intranet/bitbucket`. This more clearly states that this plugin is about "Bitbucket Server" and not "Bitbucket Cloud". Also see #225 for a cleanup regarding this.

File: src/test/java/org/jenkinsci/plugins/stashNotifier/DescriptorImplTest.java
Comment (darxriggs) [inline src/test/java/org/jenkinsci/plugins/stashNotifier/DescriptorImplTest.java]: The question is, if just verifying a `bindJSON` call is sufficient for you.
To be really sure, values should still be verified (using a `StaplerRequest` implementation/mock that implements `bindJSON` if possible). Otherwise if you for example missed a setter for a field, this would not be caught here and JCasC and UI based configuration would be broken.

File: src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java
Comment (darxriggs) [inline src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java:314]: I think this change is ok.
`Jenkins::getInstance` is now annotated with `Nonnull` and this code only runs when Jenkins is completely initialized.

File: src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java
Comment (darxriggs) [inline src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java:534]: I think this change is ok.
See comment on `getRootUrl()`.

**Review by darxriggs — CHANGES_REQUESTED** [review]
Thanks for the quick update.
I had another look at the new changes. It already looks pretty good.
See my comments for some required changes.

File: src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java
Comment (alexchiri) [reply src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java:314]: Yup, and also spotbugs doesn't give you much choice on this. Without this change and the one below, `mvn install` wouldn't complete successfully.

**Comment by alexchiri** [comment]
Thanks again for the detailed feedback and advice on this @darxriggs. Learned some things and feel I'm getting better at it. 😊 

Please let me know if anything else sticks out!

File: src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java
Comment (darxriggs) [inline src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java:681]: Now that `trimToNull()` is used in `setStashRootUrl()` this can simply return the `stashRootUrl` value (like all other getters).

File: src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java
Comment (darxriggs) [inline src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java]: You moved the field `projectKey` before the field `stashRootUrl` but the setter/getter order doesn't match this. But it doesn't really matter.

File: src/test/java/org/jenkinsci/plugins/stashNotifier/DescriptorImplTest.java
Comment (darxriggs) [inline src/test/java/org/jenkinsci/plugins/stashNotifier/DescriptorImplTest.java]: If possible try not to use PowerMockito but only Mockito. You can have a look in the internet why to avoid PowerMock.

PowerMockito is mainly required to mock constructors and static methods. That's often the case for legacy code that was written without testability in mind. At first sight that's not the case here.

File: src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java
Comment (darxriggs) [inline src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java:591]: @scaytrase I am wondring why `projectKey` is a field in `StashNotifier.DescriptorImpl`. It is not visible in the UI and was never part of `global.jelly`. Shouldn't it be removed from `StashNotifier.DescriptorImpl` and only be a field in `StashNotifier`?

**Comment by darxriggs** [comment]
Apart from the `projectKey` field issue, which is to be decided by @scaytrase, this pull request now looks good to me.
I suggest to squash all the commits into one before merging.

File: src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java
Comment (scaytrase) [reply src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java:591]: Not sure for it really, may be a copy-paste coincident. We can try to remove it, cannot image when project key should be declared on global level

**Review by scaytrase — APPROVED** [review]

File: src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java
Comment (scaytrase) [reply src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java:591]: if it's non blocking for this PR we can try it as a separate change

**Comment by scaytrase** [comment]
thank you @alexchiri ! (and @darxriggs for comments, for sure)

File: pom.xml
Comment (proski) [inline pom.xml:140]: Why was the matrix project dependency added? The commit description makes no mention of it. I don't see the matrix plugin being used in the code. I see that it's only for testing, but I still don't see those new tests.

File: src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java
Comment (proski) [inline src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java:24]: I don't think wildcards make the code better. There are plenty of tools to keep an up-to-date list of used imports.

**Review by proski — COMMENTED** [review]
It was surprising to find this large commit in the commit history with a description that makes it sound like a minor change.

File: pom.xml
Comment (darxriggs) [reply pom.xml:140]: That's a `test` scope dependency required by the new tests.

File: src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java
Comment (darxriggs) [reply src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java:24]: For me, using wildcard imports or not is a project decision. There are as always advantages and disadvantages. And nowadays tools should be able to handle both.

File: pom.xml
Comment (alexchiri) [reply pom.xml:140]: Sorry, I missed to add that. I added the Matrix project test dependency because the JenkinsRule requires it and it has been removed from core since 1.561. I added the JenkinsRule in the test that verifies the configuration as code setup, according to JCasC documentation. (import and export yaml). (this is the issue where I found more information about the NoClassDefFoundError I got while using the JenkinsRule: https://issues.jenkins-ci.org/browse/JENKINS-27826)

Comment (proski) [reply pom.xml:140]: Thank you for the explanation!

File: src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java
Comment (proski) [reply src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java:24]: I agree on all counts, I just wanted to make sure it was an intentional change and not somebody's sloppiness that everybody missed.

Comment (scaytrase) [reply src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java:24]: I'm not in general a Java developer and I adopted (ocassionaly, huh) this plugin with worse code conventions than it have today, but there is still no CSC strictly defined here today. We can adopt one, I think, if it make life easier.

**Comment by darxriggs** [comment]
@scaytrase I am looking forward for this to be released. :-)

**Comment by darxriggs** [comment]
Just as reference, there is also [JENKINS-56899](https://issues.jenkins-ci.org/browse/JENKINS-56899) tracking this.

**Pull request merged**
